modules/bitgo/src/audit.ts

```diff
@@ -0,0 +1,161 @@
+import * as base32 from 'hi-base32';
+import { decrypt, BitGoAPI } from '@bitgo/sdk-api';
+import { bitcoin } from '@bitgo/sdk-core';
+import * as stellar from 'stellar-sdk';
+import { coins, KeyCurve } from '@bitgo/statics';
+
+export type AuditKeyParams = {
+  coinName: string;
+  encryptedPrv: string;
+  walletPassphrase: string;
+} & (
+  | {
+      multiSigType: 'tss';
+      publicKey: string;
+    }
+  | {
+      multiSigType?: 'onchain';
+      publicKey?: string;
+    }
+);
+
+/**
+ * Audit a key by decrypting the encrypted private key and verifying it matches the provided public key
+ *
+ * @param coinName - The coin type
+ * @param multiSigType - Optional type of multi-signature wallet
+ * @param encryptedPrv - The encrypted private key to audit
+ * @param publicKey - The public key to verify against
+ * @param walletPassphrase - The passphrase to decrypt the encrypted private key
+ * @returns Object containing audit results
+ */
+export async function auditKey({
+  coinName,
+  publicKey,
+  encryptedPrv,
+  walletPassphrase,
+  multiSigType,
+}: AuditKeyParams): Promise<{ isValid: boolean; message?: string }> {
+  const coin = coins.get(coinName);
+  if (!coin) {
+    return { isValid: false, message: `Unsupported coin ${coinName}` };
+  }
+  try {
+    // 1. Decrypt the encrypted private key
+    const secret = decrypt(walletPassphrase, encryptedPrv);
+
+    // 2. Determine curve type based on coin and multiSigType
+    const curveType = coin.primaryKeyCurve; // default
+
+    // 3. Validate the key
+
+    switch (curveType) {
+      case KeyCurve.Secp256k1:
+        if (multiSigType === 'tss') {
+          if (publicKey.length !== 130 && publicKey.length !== 0) {
+            return { isValid: false, message: 'Incorrect TSS common keychain' };
+          }
+
+          // DKLs key chains do not have a fixed length but we know for sure they are greater than 192 in length
+          if (secret.length !== 128 && secret.length !== 192 && secret.length <= 192) {
+            return { isValid: false, message: 'Incorrect TSS keychain' };
+          }
+        } else if (multiSigType === 'onchain' && publicKey) {
```

Comment:
if publicKey isn't provided we should be returning isValid:false for multisig no?

Reply:
actually we should throw an error, saying public key needs to be provided to validate multisig key

Reply:
ok if that's the case, we need to at least validate the private key somehow, i.e. length, format, etc... as it stands the code wont validate anything if there isnt a public key provided for multisig

Reply:
i understand though length could be tricky since private keys are probably stored encoded based on how the coin likes to encode its prv's . If we can instantiate a coin instance and that coin instance has some validate key function that would be nice